core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ServiceLockSupport.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.fate.zookeeper;
+
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.apache.accumulo.core.fate.zookeeper.ServiceLock.AccumuloLockWatcher;
+import org.apache.accumulo.core.fate.zookeeper.ServiceLock.LockLossReason;
+import org.apache.accumulo.core.fate.zookeeper.ServiceLock.LockWatcher;
+import org.apache.accumulo.core.util.Halt;
+import org.apache.zookeeper.KeeperException.NoAuthException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ServiceLockSupport {
+
+  /**
+   * Lock Watcher used by Highly Available services. These are services where only instance is
+   * running at a time, but another backup service can be started that will be used if the active
+   * service instance fails and loses its lock in ZK.
+   */
+  public static class HAServiceLockWatcher implements AccumuloLockWatcher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HAServiceLockWatcher.class);
+
+    private final String serviceName;
+    private volatile boolean acquiredLock = false;
+    private volatile boolean failedToAcquireLock = false;
+
+    public HAServiceLockWatcher(String serviceName) {
```

Comment:
On the merge to `main` I would probably change this from a `String` to `ServerId.ServerType`.